lib/Remotery.c

```diff
@@ -8820,4 +8817,92 @@ RMT_API void _rmt_EndMetalSample(void)
 
 #endif // RMT_USE_METAL
 
+// Iterator
+RMT_API rmtSampleIterator _rmt_IterateChildren(rmtSample* sample)
+{
+    rmtSampleIterator iterator;
+    iterator.sample = 0;
+    iterator.initial = sample != NULL ? sample->first_child : 0;
+    return iterator;
+}
+
+RMT_API rmtBool _rmt_IterateNext(rmtSampleIterator* iter)
+{
+    if (iter->initial != NULL)
+    {
+        iter->sample = iter->initial;
+        iter->initial = 0;
+    }
+    else
+    {
+        if (iter->sample != NULL)
+            iter->sample = iter->sample->next_sibling;
+    }
+
+    return iter->sample != NULL ? RMT_TRUE : RMT_FALSE;
+}
+
+// Sample tree accessors
+RMT_API const char* _rmt_SampleTreeGetThreadName(rmtMsgSampleTree* sample_tree)
+{
+    return sample_tree->threadName;
+}
+
+RMT_API rmtSample* _rmt_SampleTreeGetRootSample(rmtMsgSampleTree* sample_tree)
+{
+    return sample_tree->rootSample;
+}
+
+// Sample accessors
+RMT_API const char* _rmt_SampleGetName(Remotery* rmt, rmtSample* sample)
+{
+    const char* name = StringTable_Find(rmt->string_table, sample->name_hash);
+    if (name == NULL)
+    {
+        return "null";
+    }
+    return name;
+}
```

Comment:
Needs the Remotery pointer, so I explicitly pass that in the callback.

Reply:
You should use `g_Remotery` in the implementation instead.

Reply:
I was thinking that too, but I was perhaps trying to remove the need to use a global variable.

The current setup in the PR was to address the initialization problem:


The problem with the user context and the remotery pointer is a catch 22:
```
// Setting up settings before creating the context
Remotery *rmt;
rmtSettings* settings = rmt_Settings();
settings->sampletree_handler = dumpTree;
settings->sampletree_context = rmt; // Not valid yet
rmt_CreateGlobalInstance(&rmt);
```

An alternative, albeit not very clean code flow either:
```
// Setup some setting variables before creating the Remotery instance...
Remotery *rmt;
rmtSettings* settings = rmt_Settings();
settings->port = port;
rmt_CreateGlobalInstance(&rmt);

// ... and set some settings after
settings->sampletree_handler = dumpTree;
settings->sampletree_context = rmt; // Not valid yet
```

But, I have no examples of other users use case, and they can for sure create their own custom contexts etc.

So I'll go ahead with removing the Remotery pointer from the callback function.

Reply:
Most of the internal code uses the global variable `g_Remotery`. The API explicitly avoids requesting it from the user. It even has a DLL API to cater for this where the use calls `rmt_SetGlobalInstance`.

Reply:
Ah, now I finally see what you mean. Fixing...